ocb/ocb.go

```diff
@@ -0,0 +1,289 @@
+// tag and nonce sizes.
+func NewOCB(block cipher.Block) (cipher.AEAD, error) {
+	return NewOCBWithNonceAndTagSize(block, defaultNonceSize, defaultTagSize)
+}
+
+// NewOCBWithNonceAndTagSize returns an OCB instance with the given block
+// cipher, nonce length, and tag length. Panics on zero nonceSize and
+// exceedingly long tag size.
+//
+// It is recommended to use at least 12 bytes as tag length.
+func NewOCBWithNonceAndTagSize(
+	block cipher.Block, nonceSize, tagSize int) (cipher.AEAD, error) {
+	if block.BlockSize() != 16 {
+		return nil, ocbError("Block cipher must have 128-bit blocks")
+	}
+	if nonceSize < 1 {
+		return nil, ocbError("Incorrect nonce length")
+	}
+	if nonceSize >= block.BlockSize() {
+		return nil, ocbError("Nonce length exceeds blocksize - 1")
+	}
+	if tagSize > block.BlockSize() {
+		return nil, ocbError("Custom tag length exceeds blocksize")
+	}
+	return &ocb{
+		block:     block,
+		tagSize:   tagSize,
+		nonceSize: nonceSize,
+		mask:      initializeMaskTable(block),
+	}, nil
+}
+
+func (o *ocb) Seal(dst, nonce, plaintext, adata []byte) []byte {
+	if len(nonce) > o.nonceSize {
+		panic("crypto/ocb: Incorrect nonce length given to OCB")
+	}
+	ret, out := byteutil.SliceForAppend(dst, len(plaintext) + o.tagSize)
+	o.crypt(enc, out, nonce, adata, plaintext)
+	return ret
+}
+
+func (o *ocb) Open(dst, nonce, ciphertext, adata []byte) ([]byte, error) {
+	if len(nonce) > o.nonceSize {
+		panic("Nonce too long for this instance")
+	}
+	if len(ciphertext) < o.tagSize {
+		return nil, ocbError("Ciphertext shorter than tag length")
+	}
+	sep := len(ciphertext) - o.tagSize
+	ret, out := byteutil.SliceForAppend(dst, len(ciphertext))
+	ciphertextData := ciphertext[:sep]
+	tag := ciphertext[sep:]
+	o.crypt(dec, out, nonce, adata, ciphertextData)
+	if subtle.ConstantTimeCompare(ret[sep:], tag) == 1 {
+		ret = ret[:sep]
+		return ret, nil
+	}
+	for i := range out {
+		out[i] = 0
+	}
+	return nil, ocbError("Tag authentication failed")
+}
+
+// On instruction enc (resp. dec), crypt is the encrypt (resp. decrypt)
+// function. It returns the resulting plain/ciphertext with the tag appended.
+func (o *ocb) crypt(instruction int, Y, nonce, adata, X []byte) []byte {
+	//
+	// Consider X as a sequence of 128-bit blocks
+	//
+	// Note: For encryption (resp. decryption), X is the plaintext (resp., the
+	// ciphertext without the tag).
+	blockSize := o.block.BlockSize()
+
+	//
+	// Nonce-dependent and per-encryption variables
+	//
+	// From RFC 7253:
+	// Nonce = num2str(TAGLEN mod 128, 7) || zeros(120 - bitlen(N)) || 1 || N
+	paddedNonce := append(make([]byte, blockSize-1-len(nonce)), 1)
+	paddedNonce = append(paddedNonce, nonce...)
+	paddedNonce[0] |= byte(((8 * o.tagSize) % (8 * blockSize)) << 1)
+	bottom := int(paddedNonce[blockSize-1] & 63)
+	// Zero out last 6 bits of paddedNonce and encrypt into Ktop
+	paddedNonce[blockSize-1] &= 192
+	Ktop := make([]byte, blockSize)
+	o.block.Encrypt(Ktop, paddedNonce)
```

Comment:
From https://mailarchive.ietf.org/arch/msg/cfrg/U4LUfK0XCEsSYNW0wHIFXVLsML0:

> OCB has a nice design feature that allows the AES operation applied to the nonce to be avoided 63 out of 64 times, if using incrementing nonces.
It is far more likely that this design feature will be tested if appendix A uses incrementing nonces (particularly for the 385 operations in each of the last tests).

With the way that OpenPGP calculates nonces as well, it seems like we could indeed reuse `Ktop` here (or even `offset`). I'm not sure whether it's worth the performance gain - perhaps you could add incrementing nonces to your benchmark function and check :) (Or just wait until we have higher-level AEAD benchmarks.) It might be worth it for small messages, which I think is common with PGP.

Reply:
`offset` depends on the original nonce (particularly, in the last 6 bits thereof), but I think we can reuse `paddedNonce`, `Ktop`, and `stretch` (therefore, only reuse `stretch` since the former variables are used only to construct it)

Reply:
Though it is wiser to reuse `Ktop` since it's lighter to store in memory and the expensive operation is the cipher block call, not the transition to `stretch`

Reply:
Just commited this.
If using incremental nonces, there is one less cipher round every 63 out of 64 ocb calls, and only storing one nonce and one block (of blockSize length) in memory.